Fixes for DRMAA when Galaxy instance isn't on a shared volume with compute nodes

vavrusa opened the pull request:
## Issues fixed
- DRMAA `workingDirectory` wasn't set correctly
- JobRunner crashed with anonymous sessions (no credentials)
- JSON job description file failed to create when strict `fopen()` semantics was followed
- Metadata calculation was always embedded in job, but it doesn't work when the Galaxy instance is detached from compute nodes
## Issues not fixed
- [Documentation page](https://wiki.galaxyproject.org/Admin/Config/Performance/Cluster) doesn't provide much help at all for DRMAA and the final email excerpt is confusing
  - `embed_metadata_in_job` added in the sample XML, but it should be referenced in the Wiki as well

vavrusa committed:
runners/drmaa: set 'workingDirectory' correctly

this fixes an issue with when the Galaxy lib (and therefore working
directory) is not shared with the compute nodes

vavrusa committed:
runners/drmaa: revert to galaxy credentials for anonymous sessions

there are no user credentials for anonymous sessions, and the runner
crashes on the log message

this patch reverts to galaxy instance credentials for anonymous sessions

vavrusa committed:
runners/drmaa: create the job description json file if not exists

vavrusa committed:
runners/drmaa: support 'embed_metadata_in_job' parameter

previously, the metadata was always computed in the job - which is a
good default, but it doesn't work if the compute nodes can't see the
Galaxy lib.

this patch adds an option for the 'drmaa' runner so it can compute
metadata externally, after the jobs are finished

natefoo commented:
Thanks for the contribution @vavrusa! I wonder if it would be best to make the fallback on submitting as the Galaxy user configurable, as some administrators may prefer to have jobs that can't be mapped to real users fail rather than be submitted by the Galaxy user.

vavrusa commented:
Makes sense, it just needs some boilerplate code so it doesn't crash with type error.

vavrusa committed:
runners/drmaa: configurable permissive mode for guest users 'allow_guests'

If enabled, anonymous and usernames that can't be mapped to the real
users will be permitted to run as a Galaxy user.

natefoo commented:
+1

martenson commented:
:+1: thanks @vavrusa for the contribution!

dannon commented:
+1
